### PR TITLE
Create the start of the profile page

### PR DESCRIPTION
**Pull request by hatched**
This currently only creates the profile page and a generic list component.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

File: Makefile [hunk 1]
```diff
@@ -165,7 +165,7 @@ $(REACT_ASSETS): $(NODE_MODULES)
 	\
 	var NoopInternalChildren = function (children) {\
 	  if (Array.isArray(children)) {\
-	    return children.map(NoopInternalChild);\
+	    return children.map(NoopInternalChildren);\
```
Comment (hatched) [inline Makefile]: Fixes for proper shallow rendering in React core

File: jujugui/static/gui/src/app/components/env-switcher/test-env-switcher.js [hunk 2]
```diff
@@ -209,7 +210,7 @@ describe('EnvSwitcher', function() {
     assert.equal(newEnv.args[0][1], envName || 'new-env-1');
     assert.equal(newEnv.args[0][2], 'admin/foo');
     assert.equal(newEnv.args[0][3], 'admin/foo');
-    assert.closeTo(newEnv.args[0][4].length, 31, 4);
+    assert.equal(newEnv.args[0][4].length > 10, true);
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/env-switcher/test-env-switcher.js]: Driveby to fix spurious test.

File: jujugui/templates/index.html.mako [hunk 3]
```diff
@@ -197,7 +197,6 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
         </header>
         <div id="content">
             <div id="shortcut-help" style="display:none"></div>
-            <div id="subapp-browser"></div>
```
Comment (hatched) [inline jujugui/templates/index.html.mako]: no longer used

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3585/
Test PASSed.

**Comment by jaycee** [comment]
QA notes:

Hitting more brought up the window; everything looked fine with two notes:

1) The user name is still the bad jem-username--env-name thing; fine for now, but I'm making a note here so we remember to fix it.

2) There needs to be better formatting off the host-ports; they run together now so the port joins with the second address.

selection_001

QA OK w/ that fixed.

**Comment by hatched** [comment]
Thanks for the QA, spacing has been fixed!

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3588/
Test FAILed.

**Comment by hatched** [comment]
cb() not called

**Comment by jaycee** [comment]
Code looks alright to me :+1: 

Someone with more react experience than me should definitely be your other reviewer though.

File: jujugui/static/gui/src/app/components/user-profile-header/user-profile-header.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+*/
+
+'use strict';
+
+YUI.add('user-profile-header', function() {
+
+  juju.components.UserProfileHeader = React.createClass({
+
+    render: function () {
+      return (<div className="user-profile-header"></div>);
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile-header/user-profile-header.js]: Is the plan to add more functionality to this component in the future?

File: jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss [hunk 5]
```diff
@@ -0,0 +1,50 @@
+.user-profile-list {
+
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss]: Don't need a space here.

File: jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        color: $light-mid-grey;
+    }
+
+    &__header-row {
+        display: flex;
+        list-style: none;
+        margin-bottom: 13px;
+    }
+
+    &__header-row :first-child {
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss]: Is this meant to be `&__header-row:first-child`?

File: jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+    &__item-row {
+        display: flex;
+        border-top: 1px solid $light-mid-grey;
+        list-style: none;
+        padding: 13px 0;
+        cursor: pointer;
+    }
+
+    &__item-row :first-child {
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss]: `&__item-row:first-child`?

File: jujugui/static/gui/src/app/components/user-profile/_user-profile.scss [hunk 8]
```diff
@@ -0,0 +1,15 @@
+.user-profile {
+
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile/_user-profile.scss]: No need for his blank line.

File: jujugui/static/gui/src/app/components/user-profile/_user-profile.scss [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    z-index: 1000;
+    top: 60px;
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile/_user-profile.scss]: You could use `$navbar-height` here (assuming that's what the space is).

File: jujugui/static/gui/src/app/components/user-profile/user-profile.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+      Makes a request of JEM or JES to fetch the users availble environments.
+
+      @method _fetchEnvironments
+    */
+    _fetchEnvironments:  function() {
+      var props = this.props;
+      var jem = props.jem;
+      if (jem) {
+        jem.listEnvironments(this._fetchEnvironmentsCallback);
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile/user-profile.js]: It might be possible for you to just pass in the listEnvironments method as a prop rather then `jem`.

File: jujugui/static/gui/src/app/components/user-profile/user-profile.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+      } else {
+        props.listEnvs(
+          'user-admin', this._fetchEnvironmentsCallback.bind(this, null));
+      }
+    },
+
+    /**
+      Callback for the JEM and JES list environments call.
+    */
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile/user-profile.js]: Lacking `@param` for `error` and `data`.

File: jujugui/static/gui/src/app/components/user-profile/user-profile.js [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    _fetchEnvironmentsCallback: function (error, data) {
+      // We need to coerce error types returned by JES vs JEM into one error.
+      var err = data.err || error;
+      if (err) {
+        console.log(err);
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile/user-profile.js]: Is it possible for this to display a "Environments could not be loaded" or something message on the page?

File: jujugui/static/gui/src/app/components/user-profile/user-profile.js [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+      var password = '';
+      var found = this.state.envList.some((env) => {
+        if (env.uuid === uuid) {
+          username = env.user;
+          password = env.password;
+          return true;
+        }
+      });
+      if (!found) {
+        console.log('No user credentials for env: ', uuid);
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile/user-profile.js]: Can we provide some feedback to the user here somehow?

File: jujugui/static/gui/src/app/components/user-profile/user-profile.js [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+      );
+    }
+
+  });
+
+}, '', {
+  requires: [
+    'panel-component',
+    'user-profile-header',
+    'user-profile-list'
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/user-profile/user-profile.js]: Need `svg-icon` here I think.

**Comment by huwshimi** [comment]
Looks good @hatched :+1: with a few little things.

File: jujugui/static/gui/src/app/components/user-profile-header/user-profile-header.js (hunk 4 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/user-profile-header/user-profile-header.js]: Yeah we're not really sure how to get the data necessary to do this so this is just a placeholder for now.

File: jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss (hunk 6 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss]: nope, I want the first child of the header row.

File: jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss (hunk 7 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss]: see above

File: jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss (hunk 5 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/user-profile-list/_user-profile-list.scss]: done

File: jujugui/static/gui/src/app/components/user-profile/_user-profile.scss (hunk 8 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/user-profile/_user-profile.scss]: done

File: jujugui/static/gui/src/app/components/user-profile/_user-profile.scss (hunk 9 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/user-profile/_user-profile.scss]: yep, done

File: jujugui/static/gui/src/app/components/user-profile/user-profile.js (hunk 10 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/user-profile/user-profile.js]: I was thinking of doing that but then I have to do the 'does jem exist' logic in the app render method as well as in here. This was less work.

File: jujugui/static/gui/src/app/components/user-profile/user-profile.js (hunk 11 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/user-profile/user-profile.js]: done

File: jujugui/static/gui/src/app/components/user-profile/user-profile.js (hunk 14 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/user-profile/user-profile.js]: done

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3591/
Test PASSed.

**Comment by hatched** [comment]
Thanks for the reviews! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge
